numba/tests/test_typedlist.py

```diff
@@ -689,6 +826,319 @@ def foo():
+
+    def test_square_bracket_builtin_from_iter_type_exception(self):
+        @njit
+        def foo():
+            l = [1, "a"]
+            return l
+        with self.assertRaises(TypingError) as raises:
+            foo()
+        # FIXME: the error message could be more specific
+        self.assertIn(
+            "Type of variable 'l' cannot be determined",
+            str(raises.exception),
+        )
+
+
+class TestConversionListToImmutableTypedList(MemoryLeakMixin, TestCase):
+
+    def test_simple_conversion(self):
+        @njit(_disable_reflected_list=True)
+        def foo(lst):
+            return lst
+        # Python list goes in and Numba immutable typed list comes out
+        received = foo([1, 2, 3])
+        expected = List()
+        [expected.append(i) for i in (1, 2, 3)]
+        # FIXME: this may fail if mutability is included in equality
+        self.assertEqual(received, expected)
+
+    def test_nested_conversion(self):
+        # nested lists are incompatible with reflected list
+        @njit(_disable_reflected_list=True)
+        def foo(lst):
+            return lst
+        a = List()
+        [a.append(i) for i in (1, 2, 3)]
+        b = List()
+        [b.append(i) for i in (4, 5, 6)]
+        expected = List()
+        expected.append(a)
+        expected.append(b)
+        received = foo([[1, 2, 3], [4, 5, 6]])
+        self.assertEqual(expected, received)
+        self.assertEqual(List, type(received))
+        self.assertEqual(List, type(received[0]))
+        self.assertEqual(List, type(received[1]))
+
+    def test_mutation_fails(self):
+        """ Test that any attempt to mutate an immutable typed list fails. """
+        def generate_function(line):
+            context = {}
+            exec_(dedent("""
+                def bar(lst):
+                    {}
+                """.format(line)), context)
+            return njit(context["bar"], _disable_reflected_list=True)
+        for line in ("lst.append(0)",
+                     "lst[0] = 0",
+                     "lst.pop()",
+                     "del lst[0]",
+                     "lst.extend((0,))",
+                     "lst.insert(0, 0)",
+                     "lst.clear()",
+                     "lst.reverse()",
+                     # FIXME: sort is missing because not implemented
+                     ):
+            with self.assertRaises(TypingError) as raises:
+                foo = generate_function(line)
+                foo([1, 2, 3])
+            self.assertIn(
+                "unable to mutate immutable typed list",
+                str(raises.exception),
+            )
+
+    def test_empty_list_raises_value_error(self):
+        @njit
+        def foo(lst):
+            return lst
+        with self.assertRaises(ValueError) as raises:
+            foo([])
+        self.assertIn(
+            "cannot compute fingerprint of empty list",
+            str(raises.exception),
+        )
```

Comment:
Is this not tested elsewhere/should the typed list equivalent be tested?

Reply:
Yes, this is missing the `_disable_reflected_list` switch, will add.